### PR TITLE
Add Fuchsia support

### PR DESCRIPTION
**Pull request by raphlinus**
Adds support for the x86_64-unknown-fuchsia target, which covers the
Fuchsia operating system.

### CODE REVIEW

**Comment by raphlinus** [comment]
@alexcrichton Note that this PR is dependent on rust-lang/libc#432 to actually work. It also requires a clang wrapper which has not yet been added to the fuchsia repository.

Comments welcome. One missing piece is aarch64 support (it's x86_64 only for now), one of the two officially supported architectures for Fuchsia. But I figure it might be useful to get feedback at this stage.

File: src/libstd/sys/unix/ext/fs.rs [hunk 1]
```diff
@@ -312,6 +312,7 @@ pub trait DirEntryExt {
 }
 
 #[stable(feature = "dir_entry_ext", since = "1.1.0")]
+#[cfg(not(target_os = "fuchsia"))]
```
Comment (alexcrichton) [inline src/libstd/sys/unix/ext/fs.rs]: We chatted on IRC, but perhaps this can be removed now?

File: src/libstd/os/fuchsia/raw.rs [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+mod arch {
+    pub use libc::{off_t, ino_t, nlink_t, blksize_t, blkcnt_t, stat, time_t};
+}
+
+#[cfg(target_arch = "s390x")]
+mod arch {
+    pub use libc::{off_t, ino_t, nlink_t, blksize_t, blkcnt_t, stat, time_t};
+}
+
+#[cfg(target_arch = "aarch64")]
```
Comment (alexcrichton) [inline src/libstd/os/fuchsia/raw.rs]: You can probably simplify a bunch of this right as only x86_64 is needed for now?

**Review by alexcrichton — COMMENTED** [review]
Looks great to me! Let's get the libc PR merged as well yeah so it can be included here.

File: src/libstd/sys/unix/ext/fs.rs (hunk 1 above)
Comment (raphlinus) [inline src/libstd/sys/unix/ext/fs.rs]: Yup, fixed.

File: src/libstd/os/fuchsia/raw.rs (hunk 2 above)
Comment (raphlinus) [inline src/libstd/os/fuchsia/raw.rs]: I chose these based on what's plausibly supported by the Magenta kernel. aarch64 is an officially supported Fuchsia target, and I'd like to get that in soon (possibly as an update to his PR). Magenta kernel also supports 32 bit targets, and I don't want to necessarily preclude Rust on those targets either.

Comment (alexcrichton) [inline src/libstd/os/fuchsia/raw.rs]: Ah never mind me then!

**Comment by alexcrichton** [comment]
Oh looks like `make tidy` is also complaining about a few locations?

**Comment by alexcrichton** [comment]
Want to update the libc submodule now as well?

**Pull request merged**